src/sagemaker_xgboost_container/checkpointing.py

```diff
@@ -0,0 +1,317 @@
+            os.makedirs(self.checkpoint_dir)
+        self.previous_checkpoints = [
+            os.path.join(self.checkpoint_dir, f)
+            for f in os.listdir(self.checkpoint_dir)]
+
+        self.thread = None
+        self.delete_queue = queue.Queue()
+
+        self.start()
+
+    def __call__(self, env):
+        """Makes the class callable since it is meant be used as a callback"""
+        return self.callback(env)
+
+    def format_path(self, iteration):
+        """Returns a file path to checkpoint given a iteration number"""
+        filename = "{}.{}".format(CHECKPOINT_FILENAME, iteration)
+        checkpoint_path = os.path.join(self.checkpoint_dir, filename)
+        return checkpoint_path
+
+    def start(self):
+        """Starts a background thread that deletes old checkpoints
+
+        To delete stale checkpoints, we use a producer-consumer pattern: we
+        start a daemon thread in the background and maintain a queue of files
+        to delete.
+        There may be a lock on the file if SageMaker is uploading the file; in
+        that case, the file is put back on the queue and we try again later.
+        When training is complete, we put SENTINEL on the queue, and when we
+        see the SENTINEL, we clean up and exit the thread.
+        """
+        def _is_uploading(path):
+            uploading = os.path.isfile(path + FILE_LOCK_SUFFIX)
+            uploaded = os.path.isfile(path + FILE_SAFE_SUFFIX)
+            return uploading and not uploaded
+
+        def _delete_once(skip_locked_files=True):
+            for iteration in iter(self.delete_queue.get, self.SENTINEL):
+                path = self.format_path(iteration)
+                # if we want to skip files that are still begin uploaded and
+                # SageMaker is uploading the file, we put the file back on the
+                # queue and try again later.
+                if skip_locked_files and _is_uploading(path):
+                    self.delete_queue.put(iteration)
+                    continue
+                try:
+                    os.remove(path)
+                except Exception:
+                    logger.debug("Failed to delete %s", path)
+                finally:
+                    self.delete_queue.task_done()
+            self.delete_queue.task_done()
+
+        def _delete_twice():
+            _delete_once(skip_locked_files=True)
+            # Here, we've reached the end of training because we place sentinel in the
+            # queue at the end of training. We put another sentinel, go through everything
+            # in the queue once again, and try to remove it anyway where there is a lock on
+            # the file or not, because the training is done. On sagemaker, this should send
+            # a delete signal to the agent so that the upload can be canceled and removed
+            # from S3.
+            self.delete_queue.put(self.SENTINEL)
+            _delete_once(skip_locked_files=False)
```

Comment:
Yes, in the normal case, that function exits immediately and doesn't do anything.

I think my attempt to reuse the same function for two different things hurt readability here. I'm going to refactor and break `_delete_once()` into two functions even if they will share some code. I hope it improves readability.

Reply:
To add, the reason I'm going through everything twice is to avoid potential file corruption when we delete a file that is still being uploaded, because the EASE team was concerned about this aspect when I talked to them. Again, the second call won't do anything in normal cases. Added code comments to address this point, and hopefully the refactor makes code easier to read.